designs/0005-static-matrices.md

```diff
@@ -0,0 +1,95 @@
+- Feature Name: static_matrices
+- Start Date: April 30th, 2020
+- RFC PR: [Example Branch](https://github.com/stan-dev/math/compare/feature/var-template)
+- Stan Issue: [#1805](https://github.com/stan-dev/math/issues/1805)
+
+# Summary
+[summary]: #summary
+
+This proposes a `static_matrix` and `static_vector` type in the Stan language which on the backend will allow for significant performance opportunities. In Stan Math this will be represented by `var<Eigen::Matrix<double, 1, 1>>` or `var<matrix_cl<double>>`. The implementation will be fully backwards compatible with current Stan code. With optimizations from the compiler the conditions for a `matrix` to be a `static_matrix` can be detected and applied to Stan programs automatically for users.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, a differentiable matrix in Stan is represented as an Eigen matrix holding a pointer to an underlying array of autodiff objects, each of those holding a pointer to the underlying autodiff object implementation. These `N*M` autodiff elements are expensive but necessary so the elements of a matrix can be assigned to without copying the entire matrix. However, in instances where the matrix is treated as a whole object such that underlying subsets of the matrix are never assigned to, we can store one autodiff object representing the entire matrix. 
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+At the language level, `static_matrix` can replace a `matrix` type if subsets of the matrix are never assigned to. That makes code such as the below illegal.
+
+```stan
+static_matrix[N, M] A = // Fill in A...
+A[10, 10] = 5;
+```
+
+Besides that they can be thought of as standard matrix.
+
+At the Stan Math level a `static_matrix` is a `var_type<Eigen::Matrix<double, -1, -1>>` with an underlying pointer to a `vari_type<Eigen::Matrix<double, -1, -1>>`*. This means that accessors for the value and adjoint of `var_type` objects can access contiguous chunks of each part of the `var_type`. Any function that accept a `var_type<T>` will support static matrices.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This feature will be added in the following stages.
+
+1. Replace `var` and `vari` with templated types. See the [example branch](https://github.com/stan-dev/math/compare/feature/var-template) here to see how this will be done. `var` and `vari` have been changed to `var_type<T>` and `vari_type<T>` with aliases for the current methods as `using var = var_type<double>` and `using vari = vari_type<double>`. These aliases allow for full backwards compatibility with existing upstream dependencies.
+
+2. Make the stack allocator conform to accept `vari_type<T>` objects.
+
+3. Additional PRs for the current arithmetic operators for `var` to accept `vari_type<T>` object types.
```

Comment:
You mean literally the arithmetic operators in C++?  Or does that also include all of our specialized stuff and scalar functions?  

If I understand, this would affect every single definition in `rev`.

Reply:
> You mean literally the arithmetic operators in C++? Or does that also include all of our specialized stuff and scalar functions?

I just mean the `+/-*` operators we have for var in this step. In my mind we would stop and do performance checks here to see what the gains are from moving to static matrices then proceed forward

> If I understand, this would affect every single definition in rev.

It matters, thanks to stuff from @andrjohns and @t4c1  a lot of functions are written in a very generic way that this wouldn't be too much. I have to look at the code that is specialized for `Eigen::Matrix<var>` rn to see if we can reuse that. Though for the probability distributions this is rather nice because we just need a specialization for `vector_seq_view`

Reply:
You mean that we could rely on the base templates in `prim`?  I don't see how that'd work with `var<matrix>` as opposed to `matrix<var>`.

All of our basic arithmetic operators are specialized for `var` types.

Reply:
> You mean that we could rely on the base templates in prim? 

Nope!

> All of our basic arithmetic operators are specialized for var types.

Indeed! This is what I'm saying

```cpp
template <typename T1, typename T2>
inline auto operator+(const var_type<T1>& a, const var_type<T2>& b) {
  using return_var = var<deduce_inner_type_t<T1, T2>>;
  return return_var{new internal::add_vv_vari<T1, T2>(a.vi_, b.vi_)};
}
```

It's sort of like that one random PR I was working on to remove the `op_vv_vari`, `op_vd_vari`, and `op_dv_vari` etc. for one `op_vari`. Then the same code can be reused for any `vari_type<T>`. You can check that implimentation of [`add_vv_vari` here](https://github.com/stan-dev/math/blob/17fa3857cf7ba3ab2779de9a14c22b56df751e77/stan/math/rev/core/operator_addition.hpp#L32)